modules/pam_faillock/tst-pam_faillock-retval.c

```diff
@@ -0,0 +1,115 @@
+/*
+ * Check pam_faillock return values.
+ */
+
+#include "test_assert.h"
+
+#include <limits.h>
+#include <stdio.h>
+#include <string.h>
+#include <unistd.h>
+#include <security/pam_appl.h>
+
+#define MODULE_NAME "pam_faillock"
+#define TEST_NAME "tst-" MODULE_NAME "-retval"
+
+static const char service_file[] = TEST_NAME ".service";
+static const char user_name[] = "root";
+static struct pam_conv conv;
+
+int
+main(void)
+{
+	pam_handle_t *pamh = NULL;
+	FILE *fp;
+	char cwd[PATH_MAX];
+	char *filename;
+	char *config_filename;
+
+	ASSERT_NE(NULL, getcwd(cwd, sizeof(cwd)));
+
+	ASSERT_LT(0,asprintf(&config_filename, "%s/my_faillock.conf", cwd));
+	ASSERT_NE(NULL, fp = fopen(config_filename, "w"));
+	ASSERT_LT(0, fprintf(fp,
+                  "deny = 2\n"
+                  "unlock_time = 5\n"
+                  "root_unlock_time = 5\n"));
+	ASSERT_EQ(0, fclose(fp));
+
+	/* root has access */
+	ASSERT_NE(NULL, fp = fopen(service_file, "w"));
+	ASSERT_LT(0, fprintf(fp, "#%%PAM-1.0\n"
+		  "auth [success=1 default=bad] %s/../pam_debug/.libs/pam_debug.so auth=success cred=success\n"
+		  "auth [default=die] %s/.libs/%s.so dir=%s authfail even_deny_root conf=%s\n"
+	          "auth sufficient %s/.libs/%s.so dir=%s authsucc even_deny_root conf=%s\n",
```

Comment:
This looks too complicated.
Why do you need all these jumps if `auth pam_debug.so auth=success cred=success` always succeeds?
Wouldn't something more simple work as well, e.g.
```
auth required %s/../pam_permit/.libs/pam_permit.so
auth required %s/.libs/%s.so authsucc even_deny_root dir=%s conf=%s
```
?

Also, I suggest testing (at least once) all module types and all 6 functions, not just `pam_authenticate`.
You can use tst-pam_sepermit-retval.c as an example.